vllm/multimodal/processing.py

```diff
@@ -107,86 +115,229 @@ def to_multi_format(data: MultiModalDataDict) -> MultiModalMultiDataDict:
+            token_ids[start_idx:end_idx],
+            # In case match_text is a special token
+            skip_special_tokens=False,
+        )
+
+        if match_text in candidate_text:
+            candidate = _TokenMatchFromTextCandidate(
+                start_idx,
+                end_idx,
+                *candidate_text.split(match_text, 1),
+            )
+
+            if candidate.distance < best_distance:
+                best_candidate = candidate
+                best_distance = candidate.distance
+
+            if best_distance == 0:
+                break
+
+    assert best_candidate is not None, dict(
+        # To facilitate debugging
+        token_ids=token_ids,
+        match_ids=match_ids,
+        left_text=left_text,
+        right_text=right_text,
+        left_idx=left_idx,
+        right_idx=right_idx,
     )
-    assert start_id == start_id_
 
-    start_id_, *candidate_ab, end_id_ = encode_no_special_tokens(
+    match_token_prefix = _cached_encode(
         tokenizer,
-        f"a{placeholder_text}b",
+        best_candidate.match_text_prefix,
+        add_special_tokens=False,
     )
-    assert start_id == start_id_ and end_id == end_id_
-
-    *candidate_b, end_id_ = encode_no_special_tokens(
+    match_token_suffix = _cached_encode(
         tokenizer,
-        f"{placeholder_text}b",
+        best_candidate.match_text_suffix,
+        add_special_tokens=False,
     )
-    assert end_id == end_id_
-
-    # Remove duplicates (need to convert to tuple to be hashable)
-    unique_candidates = {
-        tuple(c)
-        for c in [candidate_basic, candidate_a, candidate_ab, candidate_b]
-    }
 
-    # Convert back to list
-    return [list(c) for c in unique_candidates]
+    return _TokenMatchFromText(
+        start_idx=best_candidate.start_idx,
+        end_idx=best_candidate.end_idx,
+        match_prefix=match_token_prefix,
+        match_suffix=match_token_suffix,
+        match_text_prefix=best_candidate.match_text_prefix,
+        match_text_suffix=best_candidate.match_text_suffix,
+    )
 
 
-def apply_placeholders(
+def replace_by_text(
+    tokenizer: AnyTokenizer,
     token_ids: List[int],
-    placeholder_ids: List[int],
-    get_replacement_ids: Callable[[], List[int]],
-) -> Optional[PlaceholderRange]:
+    token_text: str,
+    match_text: str,
+    replacement_id: int,
+    replacement_count: int,
+) -> tuple[List[int], str, Optional[PlaceholderRange]]:
     """
-    Find the first occurrence of :code:`placeholder_ids`,
-    and replace it with the output of :code:`get_replacement_ids`.
+    Find the first occurrence of the tokenized :code:`match_text` in
+    :code:`token_ids`, and replace it with
+    :code:`[replacement_id] * replacement_count`.
 
     This function updates :code:`token_ids` in place.
     """
-    placeholder_length = len(placeholder_ids)
+    match = find_token_match_by_text(
+        tokenizer,
+        token_ids,
+        token_text,
+        match_text,
+    )
```

Comment:
I'm a little confused why we need to match by text if we already have the input sequence encoded into token_ids? Is it to deal with the situation where the image placeholder is not part of the vocab?

Reply:
Yes, the image placeholder in the text might not directly map to a single token ID.